File: packages/Server/Domain/Models/Campaign.cs
namespace Solidarity.Domain.Models;

public class Campaign : Model
{
	[MaxLength(50)] public string Title { get; set; } = string.Empty;

	public string Description { get; set; } = null!;

	public Geometry Location { get; set; } = null!;

	public DateTime TargetDate { get; set; }

	public DateTime? Completion { get; set; }

	public List<CampaignMedia> Media { get; set; } = new();

	public List<CampaignExpenditure> Expenditures { get; set; } = new();

	public long TotalExpenditure => Expenditures.Sum(e => e.TotalPrice);

	public List<CampaignPaymentMethod> ActivatedPaymentMethods { get; set; } = new();

	public int? ValidationId { get; set; }

	public Validation? Validation { get; set; } = null!;

	// public List<Milestone> Milestones { get; set; }
}

File: packages/Client/sdk/models/Campaign.ts
import { Model, Account, CampaignValidation, CampaignPaymentMethod, CampaignMedia, CampaignExpenditure, model, CampaignMediaType, CampaignAllocation } from 'sdk'
import { GeometryCollection } from 'geojson'

export enum CampaignStatus { Funding, Validation, Allocation }

@model('Campaign')
export class Campaign extends Model {
	title?: string
	description?: string
	location?: GeometryCollection
	creator?: Account
	activatedPaymentMethods = new Array<CampaignPaymentMethod>()
	media = new Array<CampaignMedia>()
	validationId?: number
	validation?: CampaignValidation
	allocation?: CampaignAllocation
	expenditures = new Array<CampaignExpenditure>()

	get status() {
		switch (true) {
			case !this.validation && !this.allocation:
				return CampaignStatus.Funding
			case !!this.validation && !this.allocation:
				return CampaignStatus.Validation
			case !!this.validation && !!this.allocation:
				return CampaignStatus.Allocation
			default:
				throw new Error('Unknown campaign status')
		}
	}

	get coverImageUri() {
		return this.media.find(m => m.type === CampaignMediaType.File)?.uri
	}

	get totalExpenditure() {
		return this.expenditures.map(e => e.totalPrice).reduce((a, acc) => a + acc, 0)
	}
}